### PR TITLE
Issue 8765, 9255 - Show informative assert errors

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
Fixes: [Issue 8765](https://issues.dlang.org/show_bug.cgi?id=8765) and [Issue 9255](https://issues.dlang.org/show_bug.cgi?id=9255)

Recreated from https://github.com/D-Programming-Language/dmd/pull/1426

Requires https://github.com/D-Programming-Language/phobos/pull/3730

### CODE REVIEW

**Comment by AndrejMitrovic** [comment]
Ok cool, these are just Phobos errors which expect a specific assert msg.

I'll write a Phobos PR fixing these first so they don't do an equality test on the `.msg` but just a `canFind`. Reason why is that I want this PR to be green when it's merged, I don't want us to have to simultaneously merge multiple PRs together as this one is known to trigger some odd x64 bug.

I'll do this the following days.

If there's a `blocked` label, please apply it, thanks!

**Comment by Geod24** [comment]
So when a message is not provided, will it still call the usual routine ? We could unify `_d_assert` and `_d_assertmsg` and provide useful error message.

File: src/expression.d [hunk 1]
```diff
@@ -7704,6 +7704,9 @@ public:
         {
             printf("AssertExp::semantic('%s')\n", toChars());
         }
+        // save expression as a message before any semantic expansion
+        char *assertExpMsg = msg ? null : mem.xstrdup(toChars());
```
Comment (ibuclaw) [inline src/expression.d]: The strdup seems redundant.

**Comment by ibuclaw** [comment]
@Geod24 this PR does nothing special around that.

File: src/expression.d [hunk 2]
```diff
@@ -7716,7 +7719,34 @@ public:
...
+                Expression assertMsg = new StringExp(Loc(), buf.extractData());
+                msg = new CatExp(Loc(), assertMsg, msg);
+            }
+            else  // compile-time value
+            {
+                msg = ensureValidMsg(sc, msg);
+                buf.printf("assert(%s) : %s", e1.toChars(),
+                           (msg.op == TOKstring) ? (cast(StringExp )msg).string : msg.toChars());
+                buf.writeByte(0);
+                msg = new StringExp(Loc(), buf.extractData());
```
Comment (yebblies) [inline src/expression.d]: `buf.extractString` is the same as `writeByte(0)` + `extractData()`

**Comment by Geod24** [comment]
@ibuclaw : I know, it was more of a "subtle" hint. Because `Assertion failure` is the most helpful message ever.

**Comment by AndrejMitrovic** [comment]
Ah, I've ran into an interesting problem. I was attempting this:

```D
assert(1 < 2, "1 is less than two"); 
=> msg = `assert(1 < 2) failed:` ~ `1 is less than 2`
```

But this is actually an implicit allocation, hence why it fails in `@nogc` functions.

We're going to have to settle with only providing the assert expression message if none is provided.

This is actually even better since there might be (and there were in druntime) tests where people explicitly tested whether an assert threw a specific error message. I'll rework this.

**Comment by Geod24** [comment]
IMO it's better this way, as changing the user's message would be slightly annoying to some. [Plus we get to kill an useless runtime hook](https://github.com/D-Programming-Language/dmd/pull/5189#issuecomment-147560160)!

**Comment by ibuclaw** [comment]
> Plus we get to kill an useless runtime hook!

``_d_assert(0)`` is not going anywhere.  ;-)

File: src/expression.d [hunk 3]
```diff
@@ -7786,6 +7789,14 @@ public:
             msg = msg.implicitCastTo(sc, Type.tchar.constOf().arrayOf());
             msg = msg.optimize(WANTvalue);
         }
+        else  // no message - use assert expression as msg
+        {
+            OutBuffer buf;
+            buf.printf("%s failed", assertExpMsg);
```
Comment (Geod24) [inline src/expression.d]: I'd rephrase as 'evaluated to false'.

Comment (Geod24) [inline src/expression.d]: Just saw the test, I understand the choice of assert now. IMO we should just provide the condition expression, rather than the full `AssertExpression`. Does that make sense ?

File: test/runnable/test42.d [hunk 4]
```diff
@@ -6009,6 +6009,35 @@ void test14510()
     fun14510(vec);
 }
 
+void test8765()
+{
+    try
+    {
+        int a = 0;
+        assert(a == 1);
```
Comment (Geod24) [inline test/runnable/test42.d]: How would it behave in the case of:
```
assert(
a
==
1
);
```

?

Comment (ibuclaw) [inline test/runnable/test42.d]: The first line of the statement by rule of thumb.

Comment (ibuclaw) [inline test/runnable/test42.d]: If you are asking what the line number will be. :-)

**Comment by adamdruppe** [comment]
hey guys this would be helpful what's up with it?

**Comment by AndrejMitrovic** [comment]
> hey guys this would be helpful what's up with it?

Last time I tested it there were bizarre failures at run-time in the test-suite. I'll try to resolve conflicts and run another test now.

**Comment by wilzbach** [comment]
Rebased to #7575
